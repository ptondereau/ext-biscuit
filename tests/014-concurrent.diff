001- bool(true)
002- bool(true)
003- bool(true)
004- bool(true)
005- bool(true)
006- bool(true)
007- bool(true)
008- bool(true)
009- bool(true)
010- bool(true)
011- bool(true)
012- bool(true)
013- bool(true)
014- bool(true)
015- bool(true)
016- bool(true)
017- bool(true)
018- bool(true)
019- bool(true)
020- bool(true)
001+ Fatal error: Uncaught BiscuitException: Invalid key data length in /home/ptondereau/Code/ext-biscuit/tests/014-concurrent.php:4
002+ Stack trace:
003+ #0 /home/ptondereau/Code/ext-biscuit/tests/014-concurrent.php(4): BiscuitKeyPair->__construct()
004+ #1 /home/ptondereau/Code/ext-biscuit/tests/014-concurrent.php(26): worker(0)
005+ #2 {main}
006+   thrown in /home/ptondereau/Code/ext-biscuit/tests/014-concurrent.php on line 4
