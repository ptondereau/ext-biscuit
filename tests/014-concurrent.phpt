--TEST--
Biscuit concurrent operations test
--SKIPIF--
<?php
if (!extension_loaded('biscuit')) die('skip');
?>
--FILE--
<?php
// Test concurrent token creation and validation
function worker($id) {
    $kp = new BiscuitKeyPair();
    $builder = new BiscuitBuilder();
    
    $builder->addFact("worker({$id})");
    $builder->addFact('right("read")');
    
    $token = $builder->build($kp);
    
    $blockBuilder = new BiscuitBlockBuilder();
    $blockBuilder->addFact("task({$id})");
    
    $token = $token->appendBlock($blockBuilder);
    
    $authorizer = $token->authorizer();
    $authorizer->addPolicy("allow if worker({$id}), task({$id})");
    
    return $authorizer->authorize();
}

// Run multiple workers
$results = [];
for ($i = 0; $i < 5; $i++) {
    $results[] = worker($i);
}

foreach ($results as $result) {
    var_dump($result);
}

// Test concurrent key pair operations
function keyWorker($id) {
    $kp = new BiscuitKeyPair();
    $serialized = $kp->serialize();
    $deserialized = BiscuitKeyPair::deserialize($serialized);
    
    $publicKey = $kp->public();
    $serializedPk = $publicKey->serialize();
    $deserializedPk = BiscuitPublicKey::deserialize($serializedPk);
    
    return true;
}

$results = [];
for ($i = 0; $i < 5; $i++) {
    $results[] = keyWorker($i);
}

foreach ($results as $result) {
    var_dump($result);
}

// Test concurrent authorizer operations
function authWorker($id) {
    $kp = new BiscuitKeyPair();
    $builder = new BiscuitBuilder();
    
    for ($i = 0; $i < 3; $i++) {
        $builder->addFact("fact{$i}({$id})");
    }
    
    $token = $builder->build($kp);
    
    $authorizers = [];
    for ($i = 0; $i < 3; $i++) {
        $authorizer = $token->authorizer();
        $authorizer->addPolicy("allow if fact{$i}({$id})");
        $authorizers[] = $authorizer->authorize();
    }
    
    return count(array_filter($authorizers)) === 3;
}

$results = [];
for ($i = 0; $i < 5; $i++) {
    $results[] = authWorker($i);
}

foreach ($results as $result) {
    var_dump($result);
}

// Test concurrent block operations
function blockWorker($id) {
    $kp = new BiscuitKeyPair();
    $builder = new BiscuitBuilder();
    $builder->addFact("base({$id})");
    
    $token = $builder->build($kp);
    
    for ($i = 0; $i < 3; $i++) {
        $blockBuilder = new BiscuitBlockBuilder();
        $blockBuilder->addFact("block{$i}({$id})");
        $token = $token->appendBlock($blockBuilder);
    }
    
    $authorizer = $token->authorizer();
    $authorizer->addPolicy("allow if base({$id})");
    
    for ($i = 0; $i < 3; $i++) {
        $authorizer->addPolicy("allow if block{$i}({$id})");
    }
    
    return $authorizer->authorize();
}

$results = [];
for ($i = 0; $i < 5; $i++) {
    $results[] = blockWorker($i);
}

foreach ($results as $result) {
    var_dump($result);
}
?>
--EXPECT--
bool(true)
bool(true)
bool(true)
bool(true)
bool(true)
bool(true)
bool(true)
bool(true)
bool(true)
bool(true)
bool(true)
bool(true)
bool(true)
bool(true)
bool(true)
bool(true)
bool(true)
bool(true)
bool(true)
bool(true) 