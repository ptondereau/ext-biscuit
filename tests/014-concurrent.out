Fatal error: Uncaught BiscuitException: Invalid key data length in /home/ptondereau/Code/ext-biscuit/tests/014-concurrent.php:4
Stack trace:
#0 /home/ptondereau/Code/ext-biscuit/tests/014-concurrent.php(4): BiscuitKeyPair->__construct()
#1 /home/ptondereau/Code/ext-biscuit/tests/014-concurrent.php(26): worker(0)
#2 {main}
  thrown in /home/ptondereau/Code/ext-biscuit/tests/014-concurrent.php on line 4